File: packages/vscode-extension/src/delivery-workflow.js
"use strict";

const { createHash, randomUUID } = require("node:crypto");

const PROJECTION_SCHEMA = "chainlesschain.delivery-flow-projection";
const ACTION_SCHEMA = "chainlesschain.delivery-action";
const ACTION_RESULT_SCHEMA = "chainlesschain.delivery-action-result";
const COMMAND_RESULT_SCHEMA = "chainlesschain.delivery-flow-command-result";
const VERSION = 1;
const DIGEST_RE = /^sha256:[0-9a-f]{64}$/i;
const ACTIONS = new Set([
  "run_gates",
  "run_preview",
  "run_review",
  "apply_fix",
  "create_pr",
  "refresh_ci",
  "publish_evidence",
  "merge",
  "archive",
]);
const STATUSES = new Set(["active", "blocked", "stopped", "completed"]);
const PHASES = new Set([
  "gates",
  "preview",
  "review",
  "fix",
  "pr",
  "ci",
  "evidence",
  "merge",
  "archive",
  "completed",
]);
const ACTION_LABELS = Object.freeze({
  run_gates: "run gates",
  run_preview: "run preview",
  run_review: "run review",
  apply_fix: "apply reviewed fix",
  create_pr: "request PR creation",
  refresh_ci: "refresh CI evidence",
  publish_evidence: "publish immutable evidence",
  merge: "request merge",
  archive: "archive verified flow",
});
const STAGES = Object.freeze([
  ["Gates", new Set(["gates"])],
  ["Preview", new Set(["preview"])],
  ["Review", new Set(["review", "fix"])],
  ["PR / CI", new Set(["pr", "ci", "evidence"])],
  ["Merge", new Set(["merge"])],
  ["Archive", new Set(["archive"])],
]);

function objectFrom(value) {
  if (value && typeof value === "object" && !Array.isArray(value)) return value;
  try {
    const parsed = JSON.parse(String(value || ""));
    return parsed && typeof parsed === "object" && !Array.isArray(parsed)
      ? parsed
      : null;
  } catch {
    return null;
  }
}

function nonNegativeInteger(value) {
  return Number.isInteger(value) && value >= 0;
}

function validFailure(value) {
  return (
    value &&
    typeof value === "object" &&
    !Array.isArray(value) &&
    typeof value.message === "string" &&
    (value.line == null || nonNegativeInteger(value.line))
  );
}

/** Fail-closed parser for the shared CLI -> IDE projection. */
function parseDeliveryProjection(value) {
  const state = objectFrom(value);
  if (
    !state ||
    state.schema !== PROJECTION_SCHEMA ||
    state.version !== VERSION ||
    state.valid !== true ||
    !String(state.flowId || "") ||
    !nonNegativeInteger(state.revision) ||
    !DIGEST_RE.test(String(state.stateDigest || "")) ||
    !STATUSES.has(state.status) ||
    !PHASES.has(state.phase) ||
    !nonNegativeInteger(state.round) ||
    !nonNegativeInteger(state.maxRounds) ||
    !nonNegativeInteger(state.noProgressRounds) ||
    !nonNegativeInteger(state.maxNoProgressRounds) ||
    !Array.isArray(state.availableActions) ||
    state.availableActions.some((action) => !ACTIONS.has(action)) ||
    new Set(state.availableActions).size !== state.availableActions.length ||
    !Array.isArray(state.failures) ||
    state.failures.some((failure) => !validFailure(failure))
  ) {
    return null;
  }
  if (state.pendingEffect != null) {
    if (
      typeof state.pendingEffect !== "object" ||
      !DIGEST_RE.test(String(state.pendingEffect.id || "")) ||
      !ACTIONS.has(state.pendingEffect.action) ||
      state.availableActions.length !== 0
    ) {
      return null;
    }
  }
  if (
    state.pr != null &&
    (typeof state.pr !== "object" ||
      !Number.isInteger(state.pr.number) ||
      state.pr.number <= 0 ||
      typeof state.pr.mergeAllowed !== "boolean")
  ) {
    return null;
  }
  if (
    state.evidence != null &&
    (typeof state.evidence !== "object" ||
      !DIGEST_RE.test(String(state.evidence.recordDigest || "")) ||
      typeof state.evidence.ready !== "boolean" ||
      (state.evidence.artifactId != null &&
        typeof state.evidence.artifactId !== "string"))
  ) {
    return null;
  }
  return state;
}

/** Fail-closed parser for a state/revision-bound external action request. */
function parseDeliveryAction(value) {
  const action = objectFrom(value);
  if (
    !action ||
    action.schema !== ACTION_SCHEMA ||
    action.version !== VERSION ||
    !String(action.flowId || "") ||
    !nonNegativeInteger(action.expectedRevision) ||
    !DIGEST_RE.test(String(action.expectedStateDigest || "")) ||
    !ACTIONS.has(action.action) ||
    !action.payload ||
    typeof action.payload !== "object" ||
    Array.isArray(action.payload)
  ) {
    return null;
  }
  return action;
}

function parseDeliveryActionResult(value) {
  const envelope = objectFrom(value);
  if (
    !envelope ||
    envelope.schema !== ACTION_RESULT_SCHEMA ||
    envelope.version !== VERSION ||
    !DIGEST_RE.test(String(envelope.effectId || "")) ||
    !envelope.result ||
    typeof envelope.result !== "object" ||
    Array.isArray(envelope.result)
  ) {
    return null;
  }
  return envelope;
}

function parseDeliveryCommandResult(value) {
  const result = objectFrom(value);
  if (
    !result ||
    result.schema !== COMMAND_RESULT_SCHEMA ||
    result.version !== VERSION ||
    !result.state ||
    typeof result.state !== "object" ||
    Array.isArray(result.state)
  ) {
    return null;
  }
  const projection = parseDeliveryProjection(result.projection);
  return projection ? { ...result, projection } : null;
}

function buildDeliveryProjectArgs(statePath) {
  if (!String(statePath || "")) throw new Error("statePath is required");
  return ["artifacts", "delivery-project", String(statePath), "--json"];
}

/** CLI argv only; the CLI owns every request/settle transition. */
function buildDeliveryStepArgs({
  statePath,
  action,
  payloadPath,
  resultPath,
  expectedRevision,
  expectedStateDigest,
  expectedEffectId,
  writeState = false,
} = {}) {
  if (!String(statePath || "")) throw new Error("statePath is required");
  if (action != null && !ACTIONS.has(action)) {
    throw new Error(`unsupported delivery action: ${action}`);
  }
  if (
    expectedRevision != null &&
    (!Number.isInteger(expectedRevision) || expectedRevision < 0)
  ) {
    throw new Error("expectedRevision must be a non-negative integer");
  }
  if (
    expectedStateDigest != null &&
    !DIGEST_RE.test(String(expectedStateDigest))
  ) {
    throw new Error("expectedStateDigest must be a full sha256 digest");
  }
  if (expectedEffectId != null && !DIGEST_RE.test(String(expectedEffectId))) {
    throw new Error("expectedEffectId must be a full sha256 digest");
  }
  const args = ["artifacts", "delivery-step", String(statePath)];
  if (action) args.push("--action", action);
  if (payloadPath) args.push("--payload-file", String(payloadPath));
  if (resultPath) args.push("--result-file", String(resultPath));
  if (expectedRevision != null) {
    args.push("--expected-revision", String(expectedRevision));
  }
  if (expectedStateDigest) {
    args.push("--expected-state-digest", String(expectedStateDigest));
  }
  if (expectedEffectId) {
    args.push("--expected-effect-id", String(expectedEffectId));
  }
  if (writeState) args.push("--write-state");
  args.push("--json");
  return args;
}

function cliText(result) {
  if (result && typeof result === "object" && "ok" in result) {
    if (result.ok !== true) {
      throw failClosed(
        result.error || "delivery CLI command failed",
        "CLI_FAILED",
      );
    }
    if (result.raw != null) return result.raw;
    if (result.json != null) return result.json;
  }
  return result;
}

function failClosed(message, code = "DELIVERY_STALE") {
  const error = new Error(message);
  error.code = code;
  return error;
}

function sameBinding(projection, token, { effect = false } = {}) {
  return Boolean(
    projection &&
    token &&
    projection.flowId === token.flowId &&
    projection.revision === token.expectedRevision &&
    projection.stateDigest === token.expectedStateDigest &&
    (!effect || projection.pendingEffect?.id === token.expectedEffectId),
  );
}

/**
 * Host-neutral controller used by the Workbench glue. It never invokes a PR,
 * CI, merge or archive provider. Its only mutation command is the exact
 * request/settle protocol exposed by `cc artifacts delivery-step`.
 */
class DeliveryWorkflowController {
  constructor({ runCli, readResultFile } = {}) {
    if (typeof runCli !== "function") throw new Error("runCli is required");
    if (typeof readResultFile !== "function") {
      throw new Error("readResultFile is required");
    }
    this._runCli = runCli;
    this._readResultFile = readResultFile;
    this._statePath = null;
    this._projection = null;
    this._confirmation = null;
    this._busy = false;
  }

  get statePath() {
    return this._statePath;
  }

  get projection() {
    return this._projection;
  }

  get busy() {
    return this._busy;
  }

  async _project(statePath) {
    const raw = cliText(
      await this._runCli(buildDeliveryProjectArgs(statePath)),
    );
    const result = parseDeliveryCommandResult(raw);
    if (!result) {
      throw failClosed(
        "invalid delivery projection from CLI",
        "INVALID_PROJECTION",
      );
    }
    return result;
  }

  async _step(args) {
    const raw = cliText(await this._runCli(args));
    const result = parseDeliveryCommandResult(raw);
    if (!result) {
      throw failClosed("invalid delivery step result from CLI", "INVALID_STEP");
    }
    return result;
  }

  async load(statePath = this._statePath) {
    if (!String(statePath || "")) throw new Error("statePath is required");
    const result = await this._project(String(statePath));
    this._statePath = String(statePath);
    this._projection = result.projection;
    this._confirmation = null;
    return this._projection;
  }

  previewRequest(action) {
    const projection = this._projection;
    if (!projection || !this._statePath) {
      throw failClosed("no delivery projection is loaded", "NOT_LOADED");
    }
    if (projection.pendingEffect) {
      throw failClosed(
        "a delivery effect is already pending",
        "EFFECT_PENDING",
      );
    }
    if (!projection.availableActions.includes(action)) {
      throw failClosed("delivery action is no longer available");
    }
    const token = Object.freeze({
      id: randomUUID(),
      kind: "request",
      statePath: this._statePath,
      flowId: projection.flowId,
      action,
      expectedRevision: projection.revision,
      expectedStateDigest: projection.stateDigest,
    });
    this._confirmation = token;
    return token;
  }

  async confirmRequest(token) {
    this._consumeConfirmation(token, "request");
    return this._exclusive(async () => {
      const latest = await this._project(token.statePath);
      this._projection = latest.projection;
      if (
        !sameBinding(latest.projection, token) ||
        latest.projection.pendingEffect ||
        !latest.projection.availableActions.includes(token.action)
      ) {
        throw failClosed("delivery request confirmation is stale");
      }
      const result = await this._step(
        buildDeliveryStepArgs({
          statePath: token.statePath,
          action: token.action,
          expectedRevision: token.expectedRevision,
          expectedStateDigest: token.expectedStateDigest,
          writeState: true,
        }),
      );
      const next = result.projection;
      if (
        next.flowId !== token.flowId ||
        next.revision <= token.expectedRevision ||
        next.pendingEffect?.action !== token.action ||
        next.availableActions.length !== 0
      ) {
        throw failClosed(
          "CLI did not return the requested pending effect",
          "INVALID_STEP",
        );
      }
      this._projection = next;
      return next;
    });
  }

  async previewSettlement(resultPath) {
    const projection = this._projection;
    const pending = projection?.pendingEffect;
    if (!projection || !pending || !this._statePath) {
      throw failClosed("no delivery effect is pending", "NO_PENDING_EFFECT");
    }
    const resultText = String(await this._readResultFile(String(resultPath)));
    const envelope = parseDeliveryActionResult(resultText);
    if (!envelope || envelope.effectId !== pending.id) {
      throw failClosed(
        "result envelope effectId does not match the pending effect",
      );
    }
    const token = Object.freeze({
      id: randomUUID(),
      kind: "settle",
      statePath: this._statePath,
      resultPath: String(resultPath),
      resultDigest: `sha256:${createHash("sha256").update(resultText).digest("hex")}`,
      flowId: projection.flowId,
      action: pending.action,
      expectedRevision: projection.revision,
      expectedStateDigest: projection.stateDigest,
      expectedEffectId: pending.id,
    });
    this._confirmation = token;
    return token;
  }

  async confirmSettlement(token) {
    this._consumeConfirmation(token, "settle");
    return this._exclusive(async () => {
      const latest = await this._project(token.statePath);
      this._projection = latest.projection;
      if (!sameBinding(latest.projection, token, { effect: true })) {
        throw failClosed("delivery settlement confirmation is stale");
      }
      const resultText = String(await this._readResultFile(token.resultPath));
      const resultDigest = `sha256:${createHash("sha256")
        .update(resultText)
        .digest("hex")}`;
      const envelope = parseDeliveryActionResult(resultText);
      if (
        resultDigest !== token.resultDigest ||
        !envelope ||
        envelope.effectId !== token.expectedEffectId
      ) {
        throw failClosed("delivery result envelope changed after preview");
      }
      const result = await this._step(
        buildDeliveryStepArgs({
          statePath: token.statePath,
          resultPath: token.resultPath,
          expectedRevision: token.expectedRevision,
          expectedStateDigest: token.expectedStateDigest,
          expectedEffectId: token.expectedEffectId,
          writeState: true,
        }),
      );
      const next = result.projection;
      if (
        next.flowId !== token.flowId ||
        next.revision <= token.expectedRevision ||
        next.pendingEffect != null
      ) {
        throw failClosed(
          "CLI did not settle the exact pending effect",
          "INVALID_STEP",
        );
      }
      this._projection = next;
      return next;
    });
  }

  _consumeConfirmation(token, kind) {
    if (!token || token !== this._confirmation || token.kind !== kind) {
      throw failClosed("delivery confirmation is stale");
    }
    this._confirmation = null;
  }

  async _exclusive(operation) {
    if (this._busy)
      throw failClosed("another delivery operation is running", "BUSY");
    this._busy = true;
    try {
      return await operation();
    } finally {
      this._busy = false;
    }
  }
}

function escapeHtml(value) {
  return String(value == null ? "" : value).replace(
    /[&<>"']/g,
    (character) =>
      ({
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        '"': "&quot;",
        "'": "&#39;",
      })[character],
  );
}

function phaseStageIndex(phase) {
  if (phase === "completed") return STAGES.length;
  return STAGES.findIndex(([, phases]) => phases.has(phase));
}

function failureLocation(failure) {
  const parts = [];
  if (failure.file) {
    parts.push(
      `${failure.file}${failure.line != null ? `:${failure.line}` : ""}`,
    );
  }
  if (failure.hunk) parts.push(failure.hunk);
  if (failure.turnId) parts.push(`turn ${failure.turnId}`);
  if (failure.toolCallId) parts.push(`tool ${failure.toolCallId}`);
  return parts.join(" · ");
}

/** Safe HTML fragment embedded in the existing Sessions Workbench. */
function renderDeliveryHtml(projection, { statePath = "", error = "" } = {}) {
  const parts = [
    '<div class="delivery-head"><strong>Delivery flow</strong><span class="delivery-controls">',
    '<button class="sec" data-delivery-command="select">Select state…</button>',
    statePath
      ? '<button class="sec" data-delivery-command="refresh">Refresh flow</button>'
      : "",
    "</span></div>",
  ];
  if (statePath) {
    parts.push(
      `<div class="muted delivery-path">${escapeHtml(statePath)}</div>`,
    );
  }
  if (error) {
    parts.push(`<div class="warn">${escapeHtml(error)}</div>`);
  }
  const state = parseDeliveryProjection(projection);
  if (!state) {
    parts.push(
      '<p class="muted">Select a CLI delivery-flow state snapshot. No delivery action is available while disconnected.</p>',
    );
    return parts.join("");
  }

  const current = phaseStageIndex(state.phase);
  parts.push('<div class="delivery-stages">');
  for (let index = 0; index < STAGES.length; index += 1) {
    const [label] = STAGES[index];
    let status =
      index < current ? "done" : index === current ? "current" : "next";
    if (state.phase === "completed") status = "done";
    if (
      (state.status === "blocked" || state.status === "stopped") &&
      index === current
    ) {
      status = "blocked";
    }
    parts.push(
      `<span class="delivery-stage ${status}">${status === "done" ? "✓" : status === "current" ? "●" : status === "blocked" ? "!" : "○"} ${label}</span>`,
    );
  }
  parts.push("</div>");
  parts.push(
    `<div><strong>${escapeHtml(state.status)} / ${escapeHtml(state.phase)}</strong> · round ${state.round}/${state.maxRounds} · no progress ${state.noProgressRounds}/${state.maxNoProgressRounds} · revision ${state.revision}</div>`,
  );
  if (state.stopReason) {
    parts.push(
      `<div class="warn">Stop reason: ${escapeHtml(state.stopReason)}</div>`,
    );
  }
  if (state.pendingEffect) {
    parts.push(
      `<div class="warn">Pending request only: ${escapeHtml(ACTION_LABELS[state.pendingEffect.action] || state.pendingEffect.action)}<br/><code>${escapeHtml(state.pendingEffect.id)}</code></div>`,
    );
  }
  if (state.failures.length) {
    parts.push("<details open><summary>Failure mapping</summary><ul>");
    for (const failure of state.failures) {
      const labels = [failure.source, failure.gateId, failure.test]
        .filter(Boolean)
        .join(" / ");
      const location = failureLocation(failure);
      parts.push(
        `<li>${labels ? `<strong>${escapeHtml(labels)}</strong>: ` : ""}${escapeHtml(failure.message)}${location ? `<div class="muted">${escapeHtml(location)}</div>` : ""}</li>`,
      );
    }
    parts.push("</ul></details>");
  }
  if (state.pr) {
    parts.push(
      `<div>PR #${state.pr.number} · head <code>${escapeHtml(state.pr.headCommitSha || "unverified")}</code> · CI <code>${escapeHtml(state.pr.ciCommitSha || "unverified")}</code> · merge ${state.pr.mergeAllowed ? "eligible" : "blocked"}</div>`,
    );
  }
  if (state.evidence) {
    parts.push(
      `<div><strong>Immutable evidence</strong>: ${state.evidence.ready ? "ready" : "not ready"} · <code>${escapeHtml(state.evidence.recordDigest)}</code>${state.evidence.artifactId ? ` · artifact ${escapeHtml(state.evidence.artifactId)}` : ""}</div>`,
    );
  }
  parts.push('<div class="delivery-actions">');
  for (const action of state.availableActions) {
    parts.push(
      `<button data-delivery-action="${escapeHtml(action)}">Request: ${escapeHtml(ACTION_LABELS[action] || action)}</button>`,
    );
  }
  if (state.pendingEffect) {
    parts.push(
      '<button data-delivery-command="settle">Settle from result JSON…</button>',
    );
  }
  parts.push(
    '</div><div class="muted">Buttons create or settle a coordinator effect through <code>cc artifacts delivery-step</code>. They never call a PR, CI, merge, or archive provider directly.</div>',
  );
  return parts.join("");
}

module.exports = {
  ACTIONS,
  ACTION_LABELS,
  parseDeliveryProjection,
  parseDeliveryAction,
  parseDeliveryActionResult,
  parseDeliveryCommandResult,
  buildDeliveryProjectArgs,
  buildDeliveryStepArgs,
  DeliveryWorkflowController,
  renderDeliveryHtml,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/DeliveryWorkflow.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

/** Pure consumer and renderer for the host-neutral CLI delivery protocol. */
public final class DeliveryWorkflow {
    private static final String PROJECTION_SCHEMA =
            "chainlesschain.delivery-flow-projection";
    private static final String ACTION_SCHEMA = "chainlesschain.delivery-action";
    private static final String ACTION_RESULT_SCHEMA =
            "chainlesschain.delivery-action-result";
    private static final String COMMAND_RESULT_SCHEMA =
            "chainlesschain.delivery-flow-command-result";
    private static final Pattern DIGEST = Pattern.compile("^sha256:[0-9a-fA-F]{64}$");
    private static final Set<String> ACTIONS = Set.of(
            "run_gates", "run_preview", "run_review", "apply_fix",
            "create_pr", "refresh_ci", "publish_evidence", "merge", "archive");
    private static final Set<String> STATUSES = Set.of(
            "active", "blocked", "stopped", "completed");
    private static final Set<String> PHASES = Set.of(
            "gates", "preview", "review", "fix", "pr", "ci", "evidence",
            "merge", "archive", "completed");
    private static final Map<String, String> ACTION_LABELS = actionLabels();
    private static final List<String> STAGE_LABELS = List.of(
            "Gates", "Preview", "Review", "PR / CI", "Merge", "Archive");

    private DeliveryWorkflow() {}

    private static Map<String, String> actionLabels() {
        Map<String, String> labels = new LinkedHashMap<>();
        labels.put("run_gates", "run gates");
        labels.put("run_preview", "run preview");
        labels.put("run_review", "run review");
        labels.put("apply_fix", "apply reviewed fix");
        labels.put("create_pr", "request PR creation");
        labels.put("refresh_ci", "refresh CI evidence");
        labels.put("publish_evidence", "publish immutable evidence");
        labels.put("merge", "request merge");
        labels.put("archive", "archive verified flow");
        return Map.copyOf(labels);
    }

    public static String actionLabel(String action) {
        return ACTION_LABELS.getOrDefault(action, String.valueOf(action));
    }

    public static boolean supportsAction(String action) {
        return ACTIONS.contains(action);
    }

    public static Map<String, Object> parseProjection(String text) {
        try {
            return validateProjection(MiniJson.parseObject(text == null ? "" : text));
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> validateProjection(Map<String, Object> state) {
        if (state == null
                || !PROJECTION_SCHEMA.equals(state.get("schema"))
                || number(state.get("version")) != 1
                || !Boolean.TRUE.equals(state.get("valid"))
                || blank(state.get("flowId"))
                || number(state.get("revision")) < 0
                || !digest(state.get("stateDigest"))
                || !STATUSES.contains(String.valueOf(state.get("status")))
                || !PHASES.contains(String.valueOf(state.get("phase")))
                || number(state.get("round")) < 0
                || number(state.get("maxRounds")) < 0
                || number(state.get("noProgressRounds")) < 0
                || number(state.get("maxNoProgressRounds")) < 0
                || !(state.get("availableActions") instanceof List)
                || !(state.get("failures") instanceof List)) {
            return null;
        }
        List<Object> available = (List<Object>) state.get("availableActions");
        Set<String> unique = new LinkedHashSet<>();
        for (Object action : available) {
            String value = String.valueOf(action);
            if (!ACTIONS.contains(value) || !unique.add(value)) return null;
        }
        for (Object rawFailure : (List<Object>) state.get("failures")) {
            if (!(rawFailure instanceof Map)) return null;
            Map<String, Object> failure = (Map<String, Object>) rawFailure;
            if (!(failure.get("message") instanceof String)) return null;
            if (failure.get("line") != null && number(failure.get("line")) < 0) return null;
        }
        Object pendingRaw = state.get("pendingEffect");
        if (pendingRaw != null) {
            if (!(pendingRaw instanceof Map)) return null;
            Map<String, Object> pending = (Map<String, Object>) pendingRaw;
            if (!digest(pending.get("id"))
                    || !ACTIONS.contains(String.valueOf(pending.get("action")))
                    || !available.isEmpty()) {
                return null;
            }
        }
        Object prRaw = state.get("pr");
        if (prRaw != null) {
            if (!(prRaw instanceof Map)) return null;
            Map<String, Object> pr = (Map<String, Object>) prRaw;
            if (number(pr.get("number")) <= 0 || !(pr.get("mergeAllowed") instanceof Boolean)) {
                return null;
            }
        }
        Object evidenceRaw = state.get("evidence");
        if (evidenceRaw != null) {
            if (!(evidenceRaw instanceof Map)) return null;
            Map<String, Object> evidence = (Map<String, Object>) evidenceRaw;
            if (!digest(evidence.get("recordDigest"))
                    || !(evidence.get("ready") instanceof Boolean)
                    || (evidence.get("artifactId") != null
                    && !(evidence.get("artifactId") instanceof String))) {
                return null;
            }
        }
        return state;
    }

    public static Map<String, Object> parseAction(String text) {
        try {
            return validateAction(MiniJson.parseObject(text == null ? "" : text));
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    public static Map<String, Object> validateAction(Map<String, Object> action) {
        if (action == null
                || !ACTION_SCHEMA.equals(action.get("schema"))
                || number(action.get("version")) != 1
                || blank(action.get("flowId"))
                || number(action.get("expectedRevision")) < 0
                || !digest(action.get("expectedStateDigest"))
                || !ACTIONS.contains(String.valueOf(action.get("action")))
                || !(action.get("payload") instanceof Map)) {
            return null;
        }
        return action;
    }

    public static Map<String, Object> parseActionResult(String text) {
        try {
            return validateActionResult(MiniJson.parseObject(text == null ? "" : text));
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    public static Map<String, Object> validateActionResult(Map<String, Object> envelope) {
        if (envelope == null
                || !ACTION_RESULT_SCHEMA.equals(envelope.get("schema"))
                || number(envelope.get("version")) != 1
                || !digest(envelope.get("effectId"))
                || !(envelope.get("result") instanceof Map)) {
            return null;
        }
        return envelope;
    }

    public static Map<String, Object> parseCommandResult(String text) {
        try {
            return validateCommandResult(MiniJson.parseObject(text == null ? "" : text));
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> validateCommandResult(Map<String, Object> result) {
        if (result == null
                || !COMMAND_RESULT_SCHEMA.equals(result.get("schema"))
                || number(result.get("version")) != 1
                || !(result.get("state") instanceof Map)
                || !(result.get("projection") instanceof Map)) {
            return null;
        }
        Map<String, Object> projection =
                validateProjection((Map<String, Object>) result.get("projection"));
        return projection == null ? null : result;
    }

    public static List<String> buildProjectArgs(String statePath) {
        if (statePath == null || statePath.trim().isEmpty()) {
            throw new IllegalArgumentException("statePath is required");
        }
        return new ArrayList<>(List.of(
                "artifacts", "delivery-project", statePath, "--json"));
    }

    /** Backwards-compatible argv builder used by the protocol conformance test. */
    public static List<String> buildStepArgs(
            String statePath, String action, String payloadPath, String resultPath) {
        return buildStepArgs(statePath, action, payloadPath, resultPath,
                null, null, null, false);
    }

    /** Build exact CAS-bound CLI argv. No provider executable is ever returned. */
    public static List<String> buildStepArgs(
            String statePath, String action, String payloadPath, String resultPath,
            Long expectedRevision, String expectedStateDigest,
            String expectedEffectId, boolean writeState) {
        if (statePath == null || statePath.trim().isEmpty()) {
            throw new IllegalArgumentException("statePath is required");
        }
        if (action != null && !ACTIONS.contains(action)) {
            throw new IllegalArgumentException("unsupported delivery action: " + action);
        }
        if (expectedRevision != null && expectedRevision < 0) {
            throw new IllegalArgumentException("expectedRevision must be non-negative");
        }
        if (expectedStateDigest != null && !digest(expectedStateDigest)) {
            throw new IllegalArgumentException("expectedStateDigest must be a full sha256 digest");
        }
        if (expectedEffectId != null && !digest(expectedEffectId)) {
            throw new IllegalArgumentException("expectedEffectId must be a full sha256 digest");
        }
        List<String> args = new ArrayList<>();
        args.add("artifacts");
        args.add("delivery-step");
        args.add(statePath);
        if (action != null) {
            args.add("--action");
            args.add(action);
        }
        if (payloadPath != null) {
            args.add("--payload-file");
            args.add(payloadPath);
        }
        if (resultPath != null) {
            args.add("--result-file");
            args.add(resultPath);
        }
        if (expectedRevision != null) {
            args.add("--expected-revision");
            args.add(String.valueOf(expectedRevision));
        }
        if (expectedStateDigest != null) {
            args.add("--expected-state-digest");
            args.add(expectedStateDigest);
        }
        if (expectedEffectId != null) {
            args.add("--expected-effect-id");
            args.add(expectedEffectId);
        }
        if (writeState) args.add("--write-state");
        args.add("--json");
        return args;
    }

    @SuppressWarnings("unchecked")
    public static List<String> availableActions(Map<String, Object> projection) {
        Map<String, Object> valid = validateProjection(projection);
        if (valid == null) return List.of();
        List<String> result = new ArrayList<>();
        for (Object action : (List<Object>) valid.get("availableActions")) {
            result.add(String.valueOf(action));
        }
        return result;
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> pendingEffect(Map<String, Object> projection) {
        Map<String, Object> valid = validateProjection(projection);
        if (valid == null || !(valid.get("pendingEffect") instanceof Map)) return null;
        return (Map<String, Object>) valid.get("pendingEffect");
    }

    /** Plain text used inside the existing Sessions Workbench Swing panel. */
    @SuppressWarnings("unchecked")
    public static String render(Map<String, Object> projection, String statePath) {
        Map<String, Object> state = validateProjection(projection);
        if (state == null) {
            return "Delivery flow\nSelect a CLI delivery-flow state snapshot. "
                    + "No delivery action is available while disconnected.";
        }
        StringBuilder out = new StringBuilder("Delivery flow\n");
        if (statePath != null && !statePath.isEmpty()) out.append(statePath).append('\n');
        int current = stageIndex(String.valueOf(state.get("phase")));
        for (int i = 0; i < STAGE_LABELS.size(); i++) {
            if (i > 0) out.append(" -> ");
            String marker = i < current || current == STAGE_LABELS.size()
                    ? "[done]" : i == current ? "[current]" : "[next]";
            out.append(marker).append(' ').append(STAGE_LABELS.get(i));
        }
        out.append('\n');
        out.append(state.get("status")).append('/').append(state.get("phase"))
                .append(" · round ").append(number(state.get("round"))).append('/')
                .append(number(state.get("maxRounds")))
                .append(" · no progress ").append(number(state.get("noProgressRounds")))
                .append('/').append(number(state.get("maxNoProgressRounds")))
                .append(" · revision ").append(number(state.get("revision"))).append('\n');
        if (state.get("stopReason") != null) {
            out.append("Stop reason: ").append(state.get("stopReason")).append('\n');
        }
        Map<String, Object> pending = pendingEffect(state);
        if (pending != null) {
            out.append("Pending request only: ")
                    .append(actionLabel(String.valueOf(pending.get("action"))))
                    .append('\n').append(pending.get("id")).append('\n');
        }
        List<Object> failures = (List<Object>) state.get("failures");
        if (!failures.isEmpty()) out.append("Failure mapping\n");
        for (Object raw : failures) {
            Map<String, Object> failure = (Map<String, Object>) raw;
            out.append("- ").append(failure.get("message"));
            String location = failureLocation(failure);
            if (!location.isEmpty()) out.append(" · ").append(location);
            out.append('\n');
        }
        if (state.get("pr") instanceof Map) {
            Map<String, Object> pr = (Map<String, Object>) state.get("pr");
            out.append("PR #").append(pr.get("number"))
                    .append(" · head ").append(orUnverified(pr.get("headCommitSha")))
                    .append(" · CI ").append(orUnverified(pr.get("ciCommitSha")))
                    .append(" · merge ")
                    .append(Boolean.TRUE.equals(pr.get("mergeAllowed")) ? "eligible" : "blocked")
                    .append('\n');
        }
        if (state.get("evidence") instanceof Map) {
            Map<String, Object> evidence = (Map<String, Object>) state.get("evidence");
            out.append("Immutable evidence: ")
                    .append(Boolean.TRUE.equals(evidence.get("ready")) ? "ready" : "not ready")
                    .append(" · ").append(evidence.get("recordDigest"));
            if (evidence.get("artifactId") != null) {
                out.append(" · artifact ").append(evidence.get("artifactId"));
            }
            out.append('\n');
        }
        List<String> actions = availableActions(state);
        if (!actions.isEmpty()) {
            out.append("Available request: ").append(actionLabel(actions.get(0))).append('\n');
        }
        out.append("Requests/settlements use cc artifacts delivery-step only; "
                + "the IDE never calls a PR, CI, merge, or archive provider directly.");
        return out.toString();
    }

    private static int stageIndex(String phase) {
        if ("completed".equals(phase)) return STAGE_LABELS.size();
        if ("gates".equals(phase)) return 0;
        if ("preview".equals(phase)) return 1;
        if ("review".equals(phase) || "fix".equals(phase)) return 2;
        if ("pr".equals(phase) || "ci".equals(phase) || "evidence".equals(phase)) return 3;
        if ("merge".equals(phase)) return 4;
        if ("archive".equals(phase)) return 5;
        return -1;
    }

    private static String failureLocation(Map<String, Object> failure) {
        List<String> parts = new ArrayList<>();
        if (!blank(failure.get("file"))) {
            String file = String.valueOf(failure.get("file"));
            if (failure.get("line") != null) file += ":" + number(failure.get("line"));
            parts.add(file);
        }
        if (!blank(failure.get("hunk"))) parts.add(String.valueOf(failure.get("hunk")));
        if (!blank(failure.get("turnId"))) parts.add(String.valueOf(failure.get("turnId")));
        if (!blank(failure.get("toolCallId"))) parts.add(String.valueOf(failure.get("toolCallId")));
        return String.join(" · ", parts);
    }

    private static String orUnverified(Object value) {
        return blank(value) ? "unverified" : String.valueOf(value);
    }

    static long number(Object value) {
        if (!(value instanceof Number)) return -1;
        if (value instanceof Byte || value instanceof Short
                || value instanceof Integer || value instanceof Long) {
            long integer = ((Number) value).longValue();
            return integer >= 0 ? integer : -1;
        }
        double decimal = ((Number) value).doubleValue();
        if (!Double.isFinite(decimal) || decimal < 0
                || decimal != Math.rint(decimal) || decimal > Long.MAX_VALUE) {
            return -1;
        }
        return (long) decimal;
    }

    static boolean digest(Object value) {
        return value != null && DIGEST.matcher(String.valueOf(value)).matches();
    }

    static boolean blank(Object value) {
        return value == null || String.valueOf(value).trim().isEmpty();
    }
}
